hal/src/rtl872x/backup_ram_hal.cpp

```diff
@@ -102,12 +105,32 @@ int hal_backup_ram_init(void) {
         g_backupRamValidMarker = HAL_BACKUP_RAM_VALID_VALUE;
     });
     if ((g_backupRamValidMarker != HAL_BACKUP_RAM_VALID_VALUE) &&
-            ((BOOT_Reason() & BIT_BOOT_DSLP_RESET_HAPPEN) || SYSTEM_FLAG(restore_backup_ram) == 1 || dctFlags.restore_backup_ram == 1)) {
-        SYSTEM_FLAG(restore_backup_ram) = 0;
+            ((BOOT_Reason() & BIT_BOOT_DSLP_RESET_HAPPEN) ||
+             (SYSTEM_FLAG(restore_backup_ram) & SYSTEM_FLAG_RESTOR_BACKUP_RAM_MASK) ||
+             (dctFlags.restore_backup_ram & SYSTEM_FLAG_RESTOR_BACKUP_RAM_MASK))) {
+        SYSTEM_FLAG(restore_backup_ram) &= ~SYSTEM_FLAG_RESTOR_BACKUP_RAM_MASK;
         dct_write_app_data(&system_flags, DCT_SYSTEM_FLAGS_OFFSET, DCT_SYSTEM_FLAGS_SIZE);
         // Woke up from deep sleep
         CHECK(hal_flash_read((uintptr_t)&platform_backup_ram_persisted_flash_start, (uint8_t*)&platform_backup_ram_all_start,
                 (size_t)&platform_backup_ram_persisted_flash_size));
+
+        if ((BOOT_Reason() & BIT_BOOT_DSLP_RESET_HAPPEN) && !(SYSTEM_FLAG(restore_backup_ram) & SYSTEM_FLAG_SESSION_DATA_STALE_MASK)) {
+            // The session info restored out of flash is valid
+            // Invalidate the session info in flash, cos it's probably stale from now on
+            auto s = new SessionPersistDataOpaque();
+            if (s) {
+                std::swap(*s, session);
```

Comment:
`session` should be restored from the backup ram flash page above (line 114), correct? 
Why do we want to erase it from flash here? Wont we immediately re-write it when `hal_backup_ram_sync` runs again? 
Isnt the original issue that the values in `session` (which is in backup RAM) will almost always be out of sync with the data written to flash during `hal_backup_ram_sync()` due to how frequently it changes? 
It doesnt seem likely that we will successfully resume a session with the current approach?

Reply:
>Why do we want to erase it from flash here?

In case that the device is reset before periodic `hal_backup_ram_sync()` is called. A session can be successfully restored for all of the sleep modes, as long as the cloud connection is disconnected gracefully. Yes, the original issue is due to the stale session data in flash, that's why in the PR we mark the session data in flash as invalid once we update the session data in RAM and mark it as valid when the `hal_backup_ram_sync()` is called.